clang/lib/CIR/Dialect/Transforms/TargetLowering/LowerFunction.cpp

```diff
@@ -329,25 +379,8 @@ mlir::Value castReturnValue(mlir::Value Src, mlir::Type Ty, LowerFunction &LF) {
   // Otherwise do coercion through memory.
   if (auto addr = findAlloca(Src.getDefiningOp())) {
     auto &rewriter = LF.getRewriter();
-    auto *ctxt = LF.LM.getMLIRContext();
-    auto ptrTy = PointerType::get(ctxt, Ty);
-    auto voidPtr = PointerType::get(ctxt, cir::VoidType::get(ctxt));
-
-    // insert alloca near the previuos one
-    auto point = rewriter.saveInsertionPoint();
-    rewriter.setInsertionPointAfter(addr);
-    auto align = LF.LM.getDataLayout().getABITypeAlign(Ty);
-    auto alignAttr = rewriter.getI64IntegerAttr(align.value());
-    auto tmp =
-        rewriter.create<AllocaOp>(Src.getLoc(), ptrTy, Ty, "tmp", alignAttr);
-    rewriter.restoreInsertionPoint(point);
-
-    auto srcVoidPtr = createBitcast(addr, voidPtr, LF);
-    auto dstVoidPtr = createBitcast(tmp, voidPtr, LF);
-    auto i64Ty = IntType::get(ctxt, 64, false);
-    auto len = rewriter.create<ConstantOp>(
-        Src.getLoc(), IntAttr::get(i64Ty, SrcSize.getFixedValue()));
-    rewriter.create<MemCpyOp>(Src.getLoc(), dstVoidPtr, srcVoidPtr, len);
+    auto tmp = createTmpAlloca(LF, Src.getLoc(), Ty);
+    createMemCpy(LF, tmp, addr, SrcSize.getFixedValue());
```

Comment:
just a refactoring - the same alloca and memcpy - all the remaining details are jidden inside `createMemCpy`